### ISSUE TITLE
Incorrect video title (sometimes)

### ISSUE THREAD

**Issue by jayathungek** [pytube/pytube#632]
This does not happen often, but sometimes ```video.title``` will return the string ```YouTube```, even if that was not the original title of the video. MWE:
```
from pytube import YouTube 
#Gangnam style video, title should be PSY - GANGNAM STYLE(강남스타일) M/V 
url = "https://www.youtube.com/watch?v=9bZkp7q19f0" 

while True:
    video = YouTube(url)
    print(video.title)
    if video.title == "YouTube":
         break
```
This should run in an infinite loop, forever printing ```PSY - GANGNAM STYLE(강남스타일) M/V```. But it always breaks within the first 20 or so iterations. Any clue?

**Comment by jayathungek** [comment]
I just had a thought - I noticed that when I navigate to a Youtube video url in my browser, the browser title bar momentarily flashes ```YouTube```  before displaying the actual video title. Could this issue be down to bad timing as a result of a slow Internet connection?

**Comment by GrautDevelopes** [comment]
https://github.com/get-pytube/pytube3/issues/71
I had this problem in a different fork. Here is my information on it.

**Comment by jayathungek** [comment]
I also tried this fix, but it seems that even scraping the ```<meta property ...``` tag still returns YouTube on occasion.

**Comment by jayathungek** [comment]
I think I've found a solution:

```
#__main__.py, line 115

if "title" not in self.player_config_args:  # type: ignore
    i_start = self.watch_html.lower().index("<title>") + len("<title>")
    i_end = self.watch_html.lower().index("</title>")
    title = self.watch_html[i_start:i_end].strip()
    while title == "YouTube":
        self.prefetch()
        i_start = self.watch_html.lower().index("<title>") + len("<title>")
        i_end = self.watch_html.lower().index("</title>")
        title = self.watch_html[i_start:i_end].strip()

    index = title.lower().rfind(" - youtube")
    title = title[:index] if index > 0 else title
    self.player_config_args["title"] = unescape(title)
```

I don't think this would be too noticeable or inefficient because this error only happens rarely, so there wouldn't be a case of the programme getting stuck in a while loop waiting for the title to change - it usually only takes 1 iteration. However, I feel like there has to be a more elegant way.

**Comment by shahraizali** [comment]
if the internet is slower it gets stuck in the loop for a couple more times before it comes out.

This seems to be working for me give this a try `__main__.py` line 114 - 117
```
    # Fix for KeyError: 'title' issue #434
    if "title" not in self.player_config_args:  # type: ignore
        title = json.loads(self.player_config_args['player_response'])['videoDetails']['title']
        self.player_config_args["title"] = unescape(title)
```

> I think I've found a solution:
> 
> ```
> #__main__.py, line 115
> 
> if "title" not in self.player_config_args:  # type: ignore
>     i_start = self.watch_html.lower().index("<title>") + len("<title>")
>     i_end = self.watch_html.lower().index("</title>")
>     title = self.watch_html[i_start:i_end].strip()
>     while title == "YouTube":
>         self.prefetch()
>         i_start = self.watch_html.lower().index("<title>") + len("<title>")
>         i_end = self.watch_html.lower().index("</title>")
>         title = self.watch_html[i_start:i_end].strip()
> 
>     index = title.lower().rfind(" - youtube")
>     title = title[:index] if index > 0 else title
>     self.player_config_args["title"] = unescape(title)
> ```
> 
> I don't think this would be too noticeable or inefficient because this error only happens rarely, so there wouldn't be a case of the programme getting stuck in a while loop waiting for the title to change - it usually only takes 1 iteration. However, I feel like there has to be a more elegant way.

**Comment by vaal12** [comment]
The issue is not about timing - it is not present in the structure indeed. For this I believe getting title from <meta> tag of the self.watch_html is more reliable.

What is working for me (see lines below "# Fix for KeyError: 'title' issue #434"):

        if self.age_restricted:
            self.player_config_args = self.vid_info
        else:
            assert self.watch_html is not None
            self.player_config_args = get_ytplayer_config(self.watch_html)["args"]
           
            # Fix for KeyError: 'title' issue #434
            if "title" not in self.player_config_args:  # type: ignore
               
                i_start = self.watch_html.lower().index("<title>") + len("<title>")
                i_end = self.watch_html.lower().index("</title>")
                title = self.watch_html[i_start:i_end].strip()
                index = title.lower().rfind(" - youtube")
                title = title[:index] if index > 0 else title
                self.player_config_args["title"] = unescape(title)
            
            #Fix for title is "Youtube" issue https://github.com/nficano/pytube/issues/632
            if "title" in self.player_config_args and \
                self.player_config_args["title"] == "YouTube" : #title is "YouTube" which is not right. Getting title from meta tag of the html
                i_start = self.watch_html.lower().index(
                    "<meta name=\"title\"")+ \
                    len("<meta name=\"title\" content=\"")
                i_end = self.watch_html.lower().index("\">", i_start)
                title = self.watch_html[i_start:i_end].strip()
                self.player_config_args["title"] = unescape(title)

        # https://github.com/nficano/pytube/issues/165
        stream_maps = ["url_encoded_fmt_stream_map"]
        if "adaptive_fmts" in self.player_config_args:
            stream_maps.append("adaptive_fmts")

**Comment by IamYVJ** [comment]
Go to the pytube's .py file
In the YouTube class, on line number 116/117, you will find an assignment to a **i_start** variable

This is the line you will see

`i_start = self.watch_html.lower().index("<title>") + len("<title>")`

Replace the above line with the following line

`i_start = self.watch_html.lower().index("<meta property=\"og:title\" content=\"") + len("<meta property=\"og:title\" content=\"")`

Save the file

This should solve the title issue

Example code to get the video title is:

`from pytube import YouTube`
`link = ""`
`yt = YouTube(link)`
`print(yt.streams[0].default_filename)`
